server/src/main/java/org/elasticsearch/index/codec/vectors/DefaultIVFVectorsWriter.java

```diff
@@ -14,104 +14,84 @@
-        int desiredClusters = (int) Math.max(Math.sqrt(floatVectorValues.size()), maxNumClusters);
-        final KMeans.Results kMeans = KMeans.cluster(
-            floatVectorValues,
-            desiredClusters,
-            false,
-            42L,
-            KMeans.KmeansInitializationMethod.PLUS_PLUS,
-            null,
-            fieldInfo.getVectorSimilarityFunction() == VectorSimilarityFunction.COSINE,
-            1,
-            15,
-            desiredClusters * 256
-        );
-        float[][] centroids = kMeans.centroids();
-        // write them
-        writeCentroids(centroids, fieldInfo, globalCentroid, centroidOutput);
-        return new OnHeapCentroidAssignmentScorer(centroids);
-    }
-
     @Override
     long[] buildAndWritePostingsLists(
         FieldInfo fieldInfo,
-        InfoStream infoStream,
-        CentroidAssignmentScorer randomCentroidScorer,
+        CentroidSupplier centroidSupplier,
         FloatVectorValues floatVectorValues,
-        IndexOutput postingsOutput
+        IndexOutput postingsOutput,
+        InfoStream infoStream,
+        CentroidAssignments centroidAssignments
     ) throws IOException {
-        IntArrayList[] clusters = new IntArrayList[randomCentroidScorer.size()];
-        for (int i = 0; i < randomCentroidScorer.size(); i++) {
-            clusters[i] = new IntArrayList(floatVectorValues.size() / randomCentroidScorer.size() / 4);
-        }
-        assignCentroids(randomCentroidScorer, floatVectorValues, clusters);
-        if (infoStream.isEnabled(IVF_VECTOR_COMPONENT)) {
-            printClusterQualityStatistics(clusters, infoStream);
-        }
+
         // write the posting lists
-        final long[] offsets = new long[randomCentroidScorer.size()];
+        final long[] offsets = new long[centroidSupplier.size()];
         OptimizedScalarQuantizer quantizer = new OptimizedScalarQuantizer(fieldInfo.getVectorSimilarityFunction());
         BinarizedFloatVectorValues binarizedByteVectorValues = new BinarizedFloatVectorValues(floatVectorValues, quantizer);
         DocIdsWriter docIdsWriter = new DocIdsWriter();
-        for (int i = 0; i < randomCentroidScorer.size(); i++) {
-            float[] centroid = randomCentroidScorer.centroid(i);
+
+        int[] assignments = centroidAssignments.assignments();
+        int[] soarAssignments = centroidAssignments.soarAssignments();
+
+        int[][] clustersForMetrics = null;
+        if (infoStream.isEnabled(IVF_VECTOR_COMPONENT)) {
+            clustersForMetrics = new int[centroidSupplier.size()][];
+        }
+
+        for (int c = 0; c < centroidSupplier.size(); c++) {
+            float[] centroid = centroidSupplier.centroid(c);
             binarizedByteVectorValues.centroid = centroid;
-            // TODO sort by distance to the centroid
-            IntArrayList cluster = clusters[i];
+
+            // TODO: remove duplicates from cluster
```

Comment:
there shouldn't be duplicates, but if there are, let's remove them immediately. 


Have you seen instances where duplicates were added? The only way I could see that happening is if the SOAR assignment and initial assignment were the same.